src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs

```diff
@@ -1278,6 +1279,76 @@ public void RegisterMissingIncludes(BlittableJsonReaderArray results, BlittableJ
             }
         }
 
+        internal void RegisterCounters(BlittableJsonReaderArray counters, bool gotAll, string[] ids)
+        {
+            if (CountersByDocId == null)
+            {
+                CountersByDocId = new Dictionary<string, (bool GotAll, Dictionary<string, long?> Values)>(StringComparer.OrdinalIgnoreCase);
+            }
+
+            if (counters == null || counters.Length == 0)
+            {
+                if (gotAll == false || ids.Length == 0)
+                    return;
+                
+                foreach (var id in ids)
+                {
+                    if (CountersByDocId.TryGetValue(id, out var cache) == false)
+                    {
+                        cache.Values = new Dictionary<string, long?>(StringComparer.OrdinalIgnoreCase);
+                        CountersByDocId.Add(id, cache);
+                    }
+                    cache.GotAll = true;
+                }
+
+                return;
+            }
+
+            foreach (BlittableJsonReaderObject counterBlittable in counters)
+            {
+                if (counterBlittable.TryGet(nameof(CounterDetail.DocumentId), out string id) == false ||
+                    counterBlittable.TryGet(nameof(CounterDetail.CounterName), out string name) == false ||
+                    counterBlittable.TryGet(nameof(CounterDetail.TotalValue), out long value) == false)
+                    continue;
+
+                if (CountersByDocId.TryGetValue(id, out var cache) == false)
+                {
+                    cache.Values = new Dictionary<string, long?>(StringComparer.OrdinalIgnoreCase);
+                    CountersByDocId.Add(id, cache);
+                }
+
+                cache.Values[name] = value;
+                if (gotAll)
+                {
+                    cache.GotAll = true;
+                }
+
+            }
+        }
+
+        internal void RegisterMissingCounters(string[] ids, string[] counters)
```

Comment:
Why have both register counters and register missing? Isn't that basically the same operation?

Reply:
Not exactly, RegisterCounters iterates over counter result and adds to cache, RegisterMissing iterates over countersToInclude and checks if in cache.
But this can be merged, yes